modules/face/src/trainFacemark.cpp

```diff
@@ -0,0 +1,348 @@
+        samples(samples_)
+        {
+        }
+        virtual void operator()( const cv::Range& range) const
+        {
+            for(size_t j = (size_t)range.start; j < (size_t)range.end; ++j){
+                (*samples)[j].shapeResiduals.resize((*samples)[j].current_shape.size());
+                for(unsigned long k=0;k<(*samples)[j].current_shape.size();k++)
+                    (*samples)[j].shapeResiduals[k]=(*samples)[j].actual_shape[k]-(*samples)[j].current_shape[k];
+            }
+        }
+    private:
+        vector<training_sample>* samples;
+};
+class getRelPixels : public ParallelLoopBody
+{
+    public:
+        getRelPixels(vector<training_sample>* samples_,FacemarkKazemiImpl& object_) :
+        samples(samples_),
+        object(object_)
+        {
+        }
+        virtual void operator()( const cv::Range& range) const
+        {
+            for (size_t j = (size_t)range.start; j < (size_t)range.end; ++j){
+                object.getRelativePixels(((*samples)[j]).current_shape,((*samples)[j]).pixel_coordinates);
+            }
+        }
+    private:
+        vector<training_sample>* samples;
+        FacemarkKazemiImpl& object;
+};
+//This function initialises the training parameters.
+bool FacemarkKazemiImpl::setTrainingParameters(String filename){
+    cout << "Reading Training Parameters " << endl;
+    FileStorage fs;
+    fs.open(filename, FileStorage::READ);
+    if (!fs.isOpened())
+    {   String error_message = "Error while opening configuration file.Aborting..";
+        CV_Error(Error::StsBadArg, error_message);
+        return false;
+    }
+    int cascade_depth_;
+    int tree_depth_;
+    int num_trees_per_cascade_level_;
+    float learning_rate_;
+    int oversampling_amount_;
+    int num_test_coordinates_;
+    float lambda_;
+    int num_test_splits_;
+    fs["cascade_depth"]>> cascade_depth_;
+    fs["tree_depth"]>> tree_depth_;
+    fs["num_trees_per_cascade_level"] >> num_trees_per_cascade_level_;
+    fs["learning_rate"] >> learning_rate_;
+    fs["oversampling_amount"] >> oversampling_amount_;
+    fs["num_test_coordinates"] >> num_test_coordinates_;
+    fs["lambda"] >> lambda_;
+    fs["num_test_splits"] >> num_test_splits_;
+    params.cascade_depth = (unsigned long)cascade_depth_;
+    params.tree_depth = (unsigned long) tree_depth_;
+    params.num_trees_per_cascade_level = (unsigned long) num_trees_per_cascade_level_;
+    params.learning_rate = (float) learning_rate_;
+    params.oversampling_amount = (unsigned long) oversampling_amount_;
+    params.num_test_coordinates = (unsigned  long) num_test_coordinates_;
+    params.lambda = (float) lambda_;
+    params.num_test_splits = (unsigned long) num_test_splits_;
+    fs.release();
+    cout<<"Parameters loaded"<<endl;
+    return true;
+}
+void FacemarkKazemiImpl::getTestCoordinates ()
+{
+    for (unsigned long i = 0; i < params.cascade_depth; ++i){
+        vector<Point2f> temp;
+        RNG rng(time(0));
```

Comment:
Probably it is better to use `cv::theRNG()` instead of initialization from "timed" seed (at least to have reproducible results).